README.md
<p align="center">
Fork of the Polybar project - a fast and easy-to-use tool for creating status bars.
</p>

**Multibar** is a fork of an existing project named Polybar and aims to help users build beautiful and highly customizable status bars
for their desktop environment, without the need of having a black belt in shell scripting.

I created this fork primarily for my personal usage as I have several improvements over the "vanilla" version and I wanted to make this maintainable.
I have changed the name since I want both this and Polybar to be installed simultaneously on my system and "_polybar2_" seemed too lame...

![default configuration screenshot](doc/_static/default.png)

My intention is to keep as much compatibility with Polybar and point out things that are different. You can use Polybar Wiki for most of the things here.

## Table of Contents

* [Introduction](#introduction)
* [Improvements over Polybar](#improvements-over-polybar)
  * [Hover actions](#hover-actions)
  * [Enhanced `internal/backlight`](#enhanced-internalbacklight)
  * [Enhanced `internal/network`](#enhanced-internalnetwork)
  * [Enhanced `internal/battery`](#enhanced-internalbattery)
* [Getting started](#getting-started)
  * [Installation](#installation)
  * [First Steps](#first-steps)
* [License](#license)
* [Signatures](#signatures)

## Introduction

The main purpose of **Multibar** is to help users create awesome status bars.
It has built-in functionality to display information about the most commonly used services.
Some of the services included so far:

- Systray icons
- Window title
- Playback controls and status display for [MPD](https://www.musicpd.org/) using [libmpdclient](https://www.musicpd.org/libs/libmpdclient/)
- [ALSA](https://www.alsa-project.org/main/index.php/Main_Page) and [PulseAudio](https://www.freedesktop.org/wiki/Software/PulseAudio/) volume controls
- Workspace and desktop panel for [bspwm](https://github.com/baskerville/bspwm) and [i3](https://github.com/i3/i3)
- Workspace module for [EWMH compliant](https://specifications.freedesktop.org/wm-spec/wm-spec-1.3.html#idm140130320786080) window managers
- Keyboard layout and indicator status
- CPU and memory load indicator
- Battery display
- Network connection details
- Backlight level
- Date and time label
- Time-based shell script execution
- Command output tailing
- User-defined menu tree
- Inter-process messaging
- And more...

[See the Polybar wiki for more details](https://github.com/polybar/polybar/wiki).

## Improvements over Polybar

### Hover actions
- Modules are now easily upgradeable (with backwards compatibility) to use new hover actions (A9 A10).
- There is `format` and `format-hover` that is evaluated when mouse is hovering over a module

#### Example usage with `internal/date`:
```ini
[module/date]
type = internal/date
interval = 2
date = [%d.%m]
time = %H:%M
date-alt = [%d.%m.%Y]
time-alt = %H:%M:%S
label = %{A9:#date.toggle:}%{A10:#date.toggle:}%time% %date%%{A9 A10}
```
results in following change when hovering over date module:

![hover off](doc/_static/date-hover.gif)

#### Example with `internal-backlight`:
```ini
[module/brightness]
type = internal/backlight
card = intel_backlight

enable-scroll = true
scroll-log = true
reverse-scroll = true
scroll-interval = 10

format = "%{T5}%{T-}<label>%{F-}  "
label =
label-hover = " %percentage%%"
```
results in following change when hovering over date module:

![hover off](doc/_static/brightness-hover.gif)

### Enhanced `internal/backlight`
This module got several new options:
- `reverse-scroll` (default off), this one is pretty self-explanatory
- `logarithmic-scroll` (default off). This changes `scroll-interval` dynamically - basically human eye is a lot more sensitive to
changes in backlight in the lower range. Meaning that perceived change from 85%->100% is basically the same as for example 1%->3%.
This option sets `scroll-interval` to 1% when brightness <5% and to 3% when brightness is <= 15%, allowing much smoother dimming
while still enabling reasonable fast scrolling for higher brightness.

### Enhanced `internal/network`
Network module got an `metric-units` option as described in [Polybar PR #3001 by @Cationiz3r](https://github.com/polybar/polybar/pull/3001)
There is also new token `%local_netmask%` that can be used to determine netmask - nice feature for someone who works with a lot of
networking.

I also added TAG `<default-gw-mark>` that gets replaced with string specified by `default-gw-mark` setting (defaults to `+`) when modules
interface is a default gw.

This module now also has all the `click-` actions, so you don't have to specify `%{A...}` for each and every label anymore.

### Enhanced `internal/battery`
This module got extra `ramp-charging` option as I like to have a special icon when charging (has a bolt inside).

Also, my laptop (DELL Precision 7760) has a feature that allows me to limit battery charge "off-hours" - so when my battery
reaches 90%, it will stop charging. Old implementation of Polybar saw this "idle" state as "discharging" with 00:00 time and quite
did not like that! I like to have orange color when discharging, so it was extra obvious that the original behavior was
not quite right. I also did not want to use `full-at` since it would be just wrong - the battery will charge up to 100%, just not
during the off-hours.

### TODO
- [ ] Global hover actions in modules - at least the ones I use ;-)
- [ ] Support for adding icons to workplace names in i3 based on what app is on workplace
- [ ] Network aliases for `internal/network` modules, so I can have home SSID displayed as  etc
- [ ] Add ICMP module, so we can monitor the internet connection (not tied to any network module)
- [ ] Sending DBUS notification on volume/brightness change
- [ ] `internal/pulseaudio` should support both sources and sinks
- [X] Mark default-gw interface in `internal/network` so I can tell what is being used as uplink


## Getting started

### Installation
If you are using **Gentoo**, you can use [my overlay](https://github.com/xoores/gentoo-overlay) that will have necessary ebuilds.

Otherwise, you will have to [build from source](https://github.com/polybar/polybar/wiki/Compiling).

### First Steps
[See the Polybar wiki for details on how to run and configure multibar](https://github.com/polybar/polybar/wiki).

## License

Both Multibar and Polybar is licensed under the MIT license. [See LICENSE for more information](https://github.com/xoores/multibar/blob/master/LICENSE).

## Signatures

All commits are signed by my GPG key.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.
Each release should have the following subsections, if entries exist, in the
given order: `Breaking`, `Build`, `Deprecated`, `Removed`, `Added`, `Changed`,
`Fixed`, `Security`.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased 1.0.0]
### Added
- Hover actions ([Polybar PR #2868 by @Astrono2](https://github.com/polybar/polybar/pull/2868))
    - Support for actions on hover via action numbers `9` and `10` ([`Polybar #1064`](https://github.com/polybar/polybar/issues/1064), [`Polybar #2868`](https://github.com/polybar/polybar/pull/2868)).
    - `custom/script`: `hover-start` and `hover-end` actions ([`Polybar #1064`](https://github.com/polybar/polybar/issues/1064), [`Polybar #2868`](https://github.com/polybar/polybar/pull/2868)).
    - Added support for TAG_LABEL (`<label>`) in ipc module  ([`Polybar #2841`](https://github.com/polybar/polybar/pull/2841))  by [@madhavpcm](https://github.com/madhavpcm).
    - Added support for format-i for each hook-i defined in ipc module ([`Polybar #2775`](https://github.com/polybar/polybar/issues/2775),  [`Polybar #2810`](https://github.com/polybar/polybar/pull/2810))  by [@madhavpcm](https://github.com/madhavpcm).
    - Added `format-hover` and `label-hover` to the base module template so any module can easily make use of this with `m_formatter->add_default_formats()`
- `internal/network`:
    - Network unit type ([Polybar PR #3001 by @Cationiz3r](https://github.com/polybar/polybar/pull/3001))
    - Added `metric-units` setting
    - Added `default-gw-mark` setting and TAG. Tag `<default-gw-mark>` is replaced with `default-gw-mark` (default `+`) when interface is default gw
    - Added `click-` actions
- `internal/backlight`: reverse-scroll, logarithmic-scroll and click-toggle ([Polybar PR #2703 by me](https://github.com/polybar/polybar/pull/2703))
    - Added `reverse-scroll` option and LEFT_CLICK action
    - Added `scroll-log` option that enables logarithmic adjustment of brightness
- `internal/battery`:
    - Added `ramp-charging`
    - Added handling of extra states (such as "Not charging")


[1.0.0]: https://github.com/xoores/multibar/

src/modules/backlight.cpp
#include "modules/backlight.hpp"

#include "drawtypes/label.hpp"
#include "drawtypes/progressbar.hpp"
#include "drawtypes/ramp.hpp"
#include "modules/meta/base.inl"
#include "utils/file.hpp"
#include "utils/math.hpp"

POLYBAR_NS

namespace modules {
  template class module<backlight_module>;

  void backlight_module::brightness_handle::filepath(const string& path) {
    if (!file_util::exists(path)) {
      throw module_error("The file '" + path + "' does not exist");
    }
    m_path = path;
  }

  float backlight_module::brightness_handle::read() const {
    return std::strtof(file_util::contents(m_path).c_str(), nullptr);
  }

  backlight_module::backlight_module(const bar_settings& bar, string name_, const config& config)
      : inotify_module<backlight_module>(bar, move(name_), config) {
    m_router->register_action(EVENT_DEC, [this]() { action_dec(); });
    m_router->register_action(EVENT_INC, [this]() { action_inc(); });
    m_router->register_action(EVENT_TOGGLE, [this]() { action_toggle(); });

    auto card = m_conf.get(name(), "card", ""s);
    if (card.empty()) {
      vector<string> backlight_card_names = file_util::list_files(string_util::replace(PATH_BACKLIGHT, "%card%", ""));
      backlight_card_names.erase(std::remove_if(backlight_card_names.begin(), backlight_card_names.end(),
                                     [&](const string& card) -> bool {
                                       auto dir = string_util::replace(PATH_BACKLIGHT, "%card%", card);
                                       return !(file_util::is_file(dir + "/actual_brightness") &&
                                                file_util::is_file(dir + "/brightness") &&
                                                file_util::is_file(dir + "/max_brightness"));
                                     }),
          backlight_card_names.end());

      if (backlight_card_names.empty()) {
        throw module_error("no viable default backlight found");
      }
      card = backlight_card_names.at(0);
      if (backlight_card_names.size() > 1) {
        m_log.warn("%s: multiple backlights found, using %s", name(), card);
      } else {
        m_log.info("%s: no backlight specified, using `%s`", name(), card);
      }
    }
    // Get flag to check if we should add scroll handlers for changing value
    m_scroll = m_conf.get(name(), "enable-scroll", m_scroll);

    // Inverted scrolling
    m_scroll_inverted = m_conf.get(name(), "reverse-scroll", m_scroll_inverted);

    m_scroll_interval = m_conf.get(name(), "scroll-interval", m_scroll_interval);

    // Use semilogarithmic setting - human eye is much more sensitive to intensity changes in the low side of the scale
    // (meaning that for example change 1%->3% is perceived as the same "step" as 90%->100%). That means we should adjust
    // our step in such a way that <=15% will result in 3% step and <5 will result in 1% step. >15% defaults to the
    // m_scroll_interval.
    m_scroll_log = m_conf.get(name(), "scroll-log", m_scroll_log);

    // Add formats and elements
    m_formatter->add_default_formats(TAG_LABEL, {TAG_LABEL, TAG_LABEL_HOVER, TAG_BAR, TAG_RAMP});

    if (m_formatter->has(TAG_LABEL)) {
      m_label = load_optional_label(m_conf, name(), TAG_LABEL, "%percentage%%");
    }

    if (m_formatter->has(TAG_LABEL_HOVER)) {
      m_label_hover = load_optional_label(m_conf, name(), TAG_LABEL_HOVER, "%percentage%%");
    }

    if (m_formatter->has(TAG_BAR)) {
      m_progressbar = load_progressbar(m_bar, m_conf, name(), TAG_BAR);
    }
    if (m_formatter->has(TAG_RAMP)) {
      m_ramp = load_ramp(m_conf, name(), TAG_RAMP);
    }

    // Build path to the sysfs folder the current/maximum brightness values are located
    m_path_backlight = string_util::replace(PATH_BACKLIGHT, "%card%", card);

    m_use_actual_brightness = m_conf.get(name(), "use-actual-brightness", m_use_actual_brightness);

    m_interval = m_conf.get<decltype(m_interval)>(name(), "poll-interval", m_use_actual_brightness? 0s : 5s);
    m_lastpoll = chrono::steady_clock::now();

    std::string brightness_type = (m_use_actual_brightness ? "actual_brightness" : "brightness");
    auto path_backlight_val = m_path_backlight + "/" + brightness_type;

    m_val.filepath(path_backlight_val);
    m_max.filepath(m_path_backlight + "/max_brightness");

    // Add inotify watch
    watch(path_backlight_val);
  }

  void backlight_module::idle() {
    if (m_interval.count() > 0) {
      /*
       * Update module after interval (if any).
       * We don't always receive inotify events for the backlight files.
       * Mainly, the `backlight` file never receive an event on modification.
       * In that case, updating at an interval is the only way to get the new value.
       */
      auto now = chrono::steady_clock::now();
      if (chrono::duration_cast<decltype(m_interval)>(now - m_lastpoll) > m_interval) {
        m_lastpoll = now;
        if (on_event({})) {
          broadcast();
        }
      }
    }

    this->inotify_module::idle();
  }

  bool backlight_module::on_event(const inotify_event& event) {
    if (event.is_valid) {
      m_log.trace("%s: on_event{filename: %s, is_dir: %s, wd: %d, cookie: %d, mask: 0x%x}", name(), event.filename,
          event.is_dir? "true" : "false", event.wd, event.cookie, event.mask);
    }

    m_max_brightness = m_max.read();
    float val = m_val.read();
    int percentage = math_util::percentage(val, m_max_brightness);

    if (m_percentage != percentage) {
      m_log.trace("%s: %d%% -> %d%% (val: %f, max: %f)", name(), m_percentage, percentage, val, m_max_brightness);
      m_percentage = percentage;

      if (m_label) {
        m_label->reset_tokens();
        m_label->replace_token("%percentage%", to_string(m_percentage));
      }

      if (m_label_hover) {
        m_label_hover->reset_tokens();
        m_label_hover->replace_token("%percentage%", to_string(m_percentage));
      }

      return true;
    }

    return false;
  }

  string backlight_module::get_output() {
    // Get the module output early so that
    // the format prefix/suffix also gets wrapped
    // with the cmd handlers
    string output{module::get_output()};

    m_builder->action(mousebtn::LEFT, *this, EVENT_TOGGLE, "");

    if (m_scroll) {
      m_builder->action(mousebtn::SCROLL_UP, *this, EVENT_INC, "");
      m_builder->action(mousebtn::SCROLL_DOWN, *this, EVENT_DEC, "");
    }

    m_builder->node(output);

    m_builder->action_close();
    m_builder->action_close();

    return m_builder->flush();
  }

  bool backlight_module::build(builder* builder, const string& tag) const {
    if (tag == TAG_BAR) {
      builder->node(m_progressbar->output(m_percentage));
    } else if (tag == TAG_RAMP) {
      builder->node(m_ramp->get_by_percentage(m_percentage));
    } else if (tag == TAG_LABEL) {
        builder->node(m_label);
    } else if (tag == TAG_LABEL_HOVER) {
        builder->node(m_label_hover);
    } else {
      return false;
    }
    return true;
  }

  int backlight_module::get_step() {
    if( !m_scroll_log ) {
      return m_scroll_interval;
    }

    if( m_percentage < 5) {
      return 1;
    }

    if( m_percentage <= 20 ) {
      return 3;
    }

    return m_scroll_interval;
  }

  void backlight_module::action_toggle() {
    set_value((m_percentage > 1 ? 1 : m_max_brightness));
  }

  void backlight_module::action_inc() {
    change_value(get_step());
  }

  void backlight_module::action_dec() {
    change_value(-get_step());
  }

  // Set absolute value
  void backlight_module::set_value(int new_value) {
    try {
      m_log.info("%s: Setting brighness value to %d", name(), new_value);
      file_util::write_contents(m_path_backlight + "/brightness", to_string(new_value));
    } catch (const exception& err) {
      m_log.err(
          "%s: Unable to change backlight value. Your system may require additional "
          "configuration. Please read the module documentation.\n(reason: %s)",
          name(), err.what());
    }
  }

  // Set relative value
  void backlight_module::change_value(int value_mod) {
    if( m_scroll_inverted ) value_mod *= -1;

    m_log.info("%s: Changing value by %d%", name(), value_mod);

    int rounded = math_util::cap<double>(m_percentage + value_mod, 0, 100.0) + 0.5;
    int value = math_util::percentage_to_value<int>(rounded, m_max_brightness);

    set_value(value);
  }
} // namespace modules

POLYBAR_NS_END
